### PR TITLE
RNDV/LANE-COUNT: removed incorrect assert

### PR DESCRIPTION
**Pull request by hoopoepg**
- there is incorrect assert which prevents fallback into
  AM mode. Removed it.

### CODE REVIEW

File: src/ucp/tag/rndv.c [hunk 1]
```diff
@@ -462,7 +462,6 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_rma_get_zcopy, (self),
     ucp_lane_index_t lane;
 
     ucp_rndv_get_lanes_count(rndv_req);
-    ucs_assert_always(rndv_req->send.rndv_get.lane_count > 0);
```
Comment (brminich) [inline src/ucp/tag/rndv.c:465]: maybe move it below the `if` handling fallback to AM?

Comment (hoopoepg) [reply src/ucp/tag/rndv.c:465]: ok

**Comment by yosefe** [comment]
can we add some gtest to repro the assert?

**Comment by hoopoepg** [comment]
no ideas how to do this... for now assert detects really issue

**Comment by yosefe** [comment]
> no ideas how to do this... for now assert detects really issue

If the as assert is needed to discover real issue then we cannot remove it.
Otherwise, need to create case where we have rndv with md_map=0 in the rkey

**Comment by hoopoepg** [comment]
added gtest

File: test/gtest/ucp/test_ucp_tag_fallback.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+UCS_TEST_P(test_ucp_tag_fallback, fallback)
+{
+    void *send_buffer;
+    void *recv_buffer;
+    ucs_status_ptr_t send_req;
+    ucs_status_ptr_t recv_req;
+    ucp_request_param_t param = {0};
+
+    /* allocate read-only pages - it force to fail ibv_reg_mr calls */
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_fallback.cc:39]: `/* allocate read-only pages - it force ibv_reg_mr() failure */`

File: test/gtest/ucp/test_ucp_tag_fallback.cc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    send_buffer = mmap(0, MSG_SIZE, PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    ASSERT_NE(send_buffer, MAP_FAILED);
+
+    recv_buffer = malloc(MSG_SIZE);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_fallback.cc:43]: can use std::string (no need to release)

File: test/gtest/ucp/test_ucp_tag_fallback.cc [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+UCS_TEST_P(test_ucp_tag_fallback, fallback)
+{
+    void *send_buffer;
+    void *recv_buffer;
+    ucs_status_ptr_t send_req;
+    ucs_status_ptr_t recv_req;
+    ucp_request_param_t param = {0};
+
+    /* allocate read-only pages - it force to fail ibv_reg_mr calls */
+    send_buffer = mmap(0, MSG_SIZE, PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_fallback.cc:40]: mmap(NULL, ... )

File: test/gtest/ucp/test_ucp_tag_fallback.cc [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return params;
+    }
+};
+
+UCS_TEST_P(test_ucp_tag_fallback, fallback)
+{
+    void *send_buffer;
+    void *recv_buffer;
+    ucs_status_ptr_t send_req;
+    ucs_status_ptr_t recv_req;
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_fallback.cc:36]: can use inline declaration for all these vars, e.g
`void *send_buffer = mmap(... `

File: test/gtest/ucp/test_ucp_tag_fallback.cc [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    wait(recv_req);
+
+    free(recv_buffer);
+    munmap(send_buffer, MSG_SIZE);
+}
+
+#define UCP_INSTANTIATE_TEST_CASE_RNDV_ZCOPY(_test_case) \
+    UCP_INSTANTIATE_TEST_CASE_TLS(_test_case, dcudx, "dc_x") \
+    UCP_INSTANTIATE_TEST_CASE_TLS(_test_case, rc,    "rc_v") \
+    UCP_INSTANTIATE_TEST_CASE_TLS(_test_case, rcx,   "rc_x")
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_fallback.cc:63]: why not all TLs?

File: test/gtest/ucp/test_ucp_tag_fallback.cc (hunk 2 above)
Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc]: reworded

File: test/gtest/ucp/test_ucp_tag_fallback.cc (hunk 4 above)
Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc]: fixed

File: test/gtest/ucp/test_ucp_tag_fallback.cc (hunk 3 above)
Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc]: used auto_ptr class

File: test/gtest/ucp/test_ucp_tag_fallback.cc [hunk 7]
```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_fallback.cc:3]: why need to create new file/class?
could just add new test to `test_ucp_tag`

File: test/gtest/ucp/test_ucp_tag_fallback.cc (hunk 6 above)
Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc:61]: test is actual for cases where registration is failed - dc/rc* are good candidates

BTW, knem transport is failed to register on R/O pages

File: test/gtest/ucp/test_ucp_tag_fallback.cc (hunk 7 above)
Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc:3]: technically it is possible, but current test uses own infrastructure, including new NBX API which could not be reused from test_ucp_tag.

@yosefe what do you think?

Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc]: merged into existing file

File: test/gtest/ucp/test_ucp_tag_fallback.cc [hunk 8]
```diff
@@ -30,30 +30,28 @@ class test_ucp_tag_fallback : public ucp_test {
...
 
-    /* allocate read-only pages - it force to fail ibv_reg_mr calls */
-    send_buffer = mmap(0, MSG_SIZE, PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    /* allocate read-only pages - it force ibv_reg_mr() failure */
+    void *send_buffer = mmap(NULL, MSG_SIZE, PROT_READ,
+                             MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
     ASSERT_NE(send_buffer, MAP_FAILED);
 
-    recv_buffer = malloc(MSG_SIZE);
+    ucs::auto_ptr<char> recv_buffer(new char[MSG_SIZE]);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_fallback.cc]: auto_ptr knows to do delete[] instead of delete?

Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc]: no, but it should not be sensitive for char[]

Comment (yosefe) [reply test/gtest/ucp/test_ucp_tag_fallback.cc]: valgrind would bark

Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag_fallback.cc]: let's wait for tests

File: test/gtest/ucp/test_ucp_tag.cc [hunk 9]
```diff
@@ -452,7 +452,7 @@ UCS_TEST_P(test_ucp_tag_fallback, fallback)
                              MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
     ASSERT_NE(send_buffer, MAP_FAILED);
 
-    ucs::auto_ptr<char> recv_buffer(new char[MSG_SIZE]);
+    ucs::auto_ptr<void, ucs::dtor::free> recv_buffer(malloc(MSG_SIZE));
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag.cc:454]: just use std::vector or std::string..

Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_tag.cc]: ok, done

**Comment by hoopoepg** [comment]
force pushed to fix commit title

**Pull request merged**